src/karma.rs

```diff
@@ -1,177 +1,291 @@
+    }
 }
 
-impl KarmaValue {
-    pub fn new(term: &str) -> KarmaValue {
-        KarmaValue {
-            term: term.to_ascii_lowercase(),
+impl Karma {
+    pub fn new(term: &str) -> Karma {
+        let now = Local::now();
+        Karma {
+            term: canonicalize_term(term),
             original_term: term.to_string(),
-            votes: KarmaChange::new(0, 0),
-            last_vote: Local::now(),
-            first_vote: Local::now()
+            votes: KarmaAmount::default(),
+            last_vote: Some(now),
+            first_vote: Some(now),
+            aliases: None,
         }
     }
 
-    pub fn vote(&mut self, karma: &Karma) {
-        self.last_vote = Local::now();
-        self.votes.up += karma.change.up;
-        self.votes.down += karma.change.down;
+    pub fn vote(&mut self, karma: &KarmaAmount) {
+        self.last_vote = Some(Local::now());
+        self.votes.up += karma.up;
+        self.votes.down += karma.down;
     }
 
-    pub fn from_json(data: json::Json) -> Result<KarmaValue, Box<::std::error::Error>> {
-        macro_rules! get_key {
-            ($o:expr, $s:expr, $i_is:ident, $i_as:ident) => (match $o.get($s) {
-                Some(m) if m.$i_is() => m.$i_as().unwrap(),
-                _ => return Err(From::from(KarmaError::new(&format!("No value or invalid value for key '{}'", $s)[..])))
-            });
+    fn from_response(r: &Response) -> Result<Karma, Box<dyn std::error::Error>> {
+        match r.get_str("value") {
+            Some(s) => match serde_json::de::from_str(s) {
+                Ok(karma) => Ok(karma),
+                Err(err) => Err(err.into()),
+            },
+            None => Err(KarmaError::new("no value found in response").into()),
         }
+    }
 
-        if data.is_object() {
-            let obj = data.as_object().unwrap();
-            let term = get_key!(obj, "term", is_string, as_string);
-            let first_vote_str = get_key!(obj, "first_vote", is_string, as_string);
-            let last_vote_str = get_key!(obj, "last_vote", is_string, as_string);
-            let votes = get_key!(obj, "votes", is_object, as_object);
-
-            let upvotes = match votes.get("up") {
-                Some(m) if m.is_u64() => m.as_u64().unwrap(),
-                Some(m) if m.is_string() && m.as_string().unwrap().parse::<u64>().is_ok() => m.as_string().unwrap().parse().unwrap(),
-                _ => return Err(From::from(KarmaError::new("No value or invalid value for key 'up'")))
-            };
+    pub fn get_from_dazeus(
+        dazeus: &dyn DaZeusClient,
+        scope: Scope,
+        term: &str,
+    ) -> Result<Karma, Box<dyn std::error::Error>> {
+        let property = format!("{}{}", STORE_PREFIX, canonicalize_term(term));
+        let json = dazeus.get_property(&property[..], scope);
+        let mut karma = Karma::from_response(&json);
+        if let Ok(ref mut k) = karma {
+            k.original_term = term.to_owned();
+        }
+        karma
```

Comment:
I think it's clearer to handle the error case first:
```suggestion
        let mut karma = Karma::from_response(&json)?;
        karma.original_term = term.to_owned();
        Ok(karma)
```

Reply:
Also, `from_response` should probably be inlined here since it isn't used anywhere else.

Reply:
> 
> 
> I think it's clearer to handle the error case first:

I'll fix this.

Reply:
> 
> 
> Also, `from_response` should probably be inlined here since it isn't used anywhere else.

Nack. This is similar how it looked first. In particular, I'd like to keep these functions split for (future) testing.